x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -1615,6 +1618,23 @@ private LogicalPlan rule(Eval eval) {
         }
     }
 
+    /**
+     * Replace type converting eval with aliasing eval when type change does not occur.
+     * A following ReplaceAliasingEvalWithProject will effectively convert ReferenceAttribute into FieldAttribute,
+     * something very useful in local physical planning.
+     */
+    static class ReplaceTypeConvertingEvalWithAliasingEval extends OptimizerRules.OptimizerRule<Eval> {
+        @Override
+        protected LogicalPlan rule(Eval eval) {
+            return eval.transformExpressionsOnly(AbstractConvertFunction.class, convert -> {
+                if (convert.field() instanceof FieldAttribute fa && fa.dataType() == convert.dataType()) {
```

Comment:
Hmm, why only in case of field attrs? Wouldn't this more general optimization also work?
```suggestion
                if (fa.dataType() == convert.dataType()) {
```

Reply:
I only changed FieldAttribute because my primary goal was to enable lucene optimizations done by physical planning, and ReferenceAttributes that hide FieldAttributes were the main issue. Expanding to a wider scope could make sense, but also does not change performance in any way at all. The current PR makes one query 75x faster! I was already mildly concerned that my changes would have some subtle unforeseen side-effects, so did not want to widen the scope beyond what I actually needed.